Generate "process synched data" code

beidson opened the pull request:
#### 376f4eaf9750f0df5b0f91ab0dad9e81fe217c37
<pre>
Generate &quot;process synched data&quot; code
<a href="https://bugs.webkit.org/show_bug.cgi?id=283178">https://bugs.webkit.org/show_bug.cgi?id=283178</a>
&lt;<a href="https://rdar.apple.com/problem/139974832">rdar://problem/139974832</a>&gt;

Reviewed by Charlie Wolfe.

With site isolation, there&apos;s a lot of bits of data that are generated in one WebContent process that
other WebContent processes need to be aware of.

We&apos;d started down a somewhat slow road of writing messages for each case, ad hoc.
Keeping with that approach seems undesirable.

This patch adds a new `ProcessSyncData.in` file where you declare a property that needs to be synced.
It also adds a script that processes that file.

The script generates `ProcessSyncClient` with one method per data to broadcast the change.
It also generates `ProcessSyncData` code to wrap the data type being broadcast as well as generate its
serializers for message passing.

To add a new synced data type is now as simple as:
1 - Adding an entry for it in `ProcessSyncData.in`
2 - Make a one line call to the new `ProcessSyncClient` method where the data change is generated
3 - Handle the new switch case in `Page::updateProcessSyncData` to receive the data change broadcast

I fully expect this mechanism will grow to become much more flexible (e.g. targeting specific frames)
but for now most of the complexity was getting all of the scripts and build system stuff figured out;
Future expansion of the mechanism should be much easier.

The behavior change of moving &quot;MainFrameURLChange&quot; over to this new mechanism is covered by an API test:
SiteIsolation.MainFrameURLAfterFragmentNavigation

* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:

* Source/WebCore/Scripts/generate-process-sync-data.py: Added.
(SyncedData):
(SyncedData.__init__):
(parse_process_sync_data):
(generate_process_sync_client_header):
(generate_process_sync_client_impl):
(sorted_qualified_types):
(generate_process_sync_data_header):
(generate_process_sync_data_header.ProcessSyncDataType):
(generate_process_sync_data_serialiation_in):
(generate_process_sync_data_serialiation_in.WebCore):
(main):

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateProcessSyncData):
* Source/WebCore/page/Page.h:

* Source/WebCore/page/ProcessSyncClient.h: Removed.

* Source/WebCore/page/ProcessSyncData.in: Added.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::broadcastProcessSyncData):
(WebKit::WebPageProxy::broadcastMainFrameURLChangeToOtherProcesses): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:

* Source/WebKit/WebProcess/WebCoreSupport/WebProcessSyncClient.cpp:
(WebKit::WebProcessSyncClient::broadcastProcessSyncDataToOtherProcesses):
(WebKit::WebProcessSyncClient::broadcastMainFrameURLChangeToOtherProcesses): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebProcessSyncClient.h:

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::processSyncDataChangedInAnotherProcess):
(WebKit::WebPage::mainFrameURLChangedInAnotherProcess): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/286671@main">https://commits.webkit.org/286671@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b25d1f83dbd9f9de203ff7e9ad2004708615a13c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76704 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55739 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29610 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81236 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27980 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78821 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64881 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4032 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/60126 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18210 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79771 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/50057 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65858 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40448 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47457 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23352 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26304 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68577 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23682 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82681 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4080 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2695 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68394 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4233 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65830 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67645 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16873 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11629 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9707 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4027 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6836 "Found 1 new failure in ByteLengthQueuingStrategyBuiltins.h") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4050 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7480 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5808 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebCore/Scripts/generate-process-sync-data.py:
Isn’t `Source/WebKit/Scripts/` better place for this to be? Along with the other IPC related code generators.

charliewolfe reviewed (approved):
Other code generators have tests with expected generated output in `Source/WebKit/Scripts/tests/` , maybe we should have the same for this one.

beidson committed:
Generate "process synched data" code
https://bugs.webkit.org/show_bug.cgi?id=283178
<rdar://problem/139974832>

Reviewed by Charlie Wolfe.

With site isolation, there's a lot of bits of data that are generated in one WebContent process that
other WebContent processes need to be aware of.

We'd started down a somewhat slow road of writing messages for each case, ad hoc.
Keeping with that approach seems undesirable.

This patch adds a new `ProcessSyncData.in` file where you declare a property that needs to be synced.
It also adds a script that processes that file.

The script generates `ProcessSyncClient` with one method per data to broadcast the change.
It also generates `ProcessSyncData` code to wrap the data type being broadcast as well as generate its
serializers for message passing.

To add a new synced data type is now as simple as:
1 - Adding an entry for it in `ProcessSyncData.in`
2 - Make a one line call to the new `ProcessSyncClient` method where the data change is generated
3 - Handle the new switch case in `Page::updateProcessSyncData` to receive the data change broadcast

I fully expect this mechanism will grow to become much more flexible (e.g. targeting specific frames)
but for now most of the complexity was getting all of the scripts and build system stuff figured out;
Future expansion of the mechanism should be much easier.

The behavior change of moving "MainFrameURLChange" over to this new mechanism is covered by an API test:
SiteIsolation.MainFrameURLAfterFragmentNavigation

* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:

* Source/WebCore/Scripts/generate-process-sync-data.py: Added.
(SyncedData):
(SyncedData.__init__):
(parse_process_sync_data):
(generate_process_sync_client_header):
(generate_process_sync_client_impl):
(sorted_qualified_types):
(generate_process_sync_data_header):
(generate_process_sync_data_header.ProcessSyncDataType):
(generate_process_sync_data_serialiation_in):
(generate_process_sync_data_serialiation_in.WebCore):
(main):

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateProcessSyncData):
* Source/WebCore/page/Page.h:

* Source/WebCore/page/ProcessSyncClient.h: Removed.

* Source/WebCore/page/ProcessSyncData.in: Added.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::broadcastProcessSyncData):
(WebKit::WebPageProxy::broadcastMainFrameURLChangeToOtherProcesses): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:

* Source/WebKit/WebProcess/WebCoreSupport/WebProcessSyncClient.cpp:
(WebKit::WebProcessSyncClient::broadcastProcessSyncDataToOtherProcesses):
(WebKit::WebProcessSyncClient::broadcastMainFrameURLChangeToOtherProcesses): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebProcessSyncClient.h:

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::processSyncDataChangedInAnotherProcess):
(WebKit::WebPage::mainFrameURLChangedInAnotherProcess): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/286671@main

burg commented on Source/WebCore/CMakeLists.txt:
```diff
@@ -2528,6 +2528,17 @@ add_custom_command(
+    COMMAND ${PYTHON_EXECUTABLE} ${WEBCORE_DIR}/Scripts/generate-process-sync-data.py ${WEBCORE_DIR}/page/ProcessSyncData.in ${WebCore_DERIVED_SOURCES_DIR}
+    VERBATIM)
+list(APPEND WebCore_SOURCES ${WebCore_DERIVED_SOURCES_DIR}/ProcessSyncClient.cpp)
+list(APPEND WebCore_SOURCES ${WebCore_DERIVED_SOURCES_DIR}/ProcessSyncClient.h)
```
Nit: is it necessary to add header files to WebCore_SOURCES? I would have expected them to be added to private headers instead, if anything.

burg commented on Source/WebCore/Scripts/generate-process-sync-data.py:
```diff
@@ -0,0 +1,302 @@
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+"""
```
Nit: I recommend you add "DO NOT EDIT; this file was generated by [scriptname]" to the generated output.

burg reviewed (commented):
Cool.
